cdap-data-fabric/src/main/java/co/cask/cdap/store/DefaultNamespaceStore.java

```
@@ -0,0 +1,158 @@
+              dsFramework, APP_META_INSTANCE_ID, "table",
+              DatasetProperties.EMPTY, DatasetDefinition.NO_ARGUMENTS, null);
+            return new NamespaceMetadataStore(mdsTable);
+          } catch (DatasetManagementException | IOException e) {
+            throw Throwables.propagate(e);
+          }
+        }
+      }
+    );
+    this.appsTx = new Supplier<TransactionExecutor>() {
+      @Override
+      public TransactionExecutor get() {
+        return txExecutorFactory.createExecutor(ImmutableList.of((TransactionAware) apps.get()));
+      }
+    };
+  }
+
+  @Override
+  @Nullable
+  public NamespaceMeta create(final NamespaceMeta metadata) {
+    Preconditions.checkArgument(metadata != null, "Namespace metadata cannot be null.");
+    return appsTx.get().executeUnchecked(
+      new TransactionExecutor.Function<NamespaceMetadataStore, NamespaceMeta>() {
+        @Override
+        public NamespaceMeta apply(NamespaceMetadataStore mds) throws Exception {
+          Id.Namespace namespace = Id.Namespace.from(metadata.getName());
+          NamespaceMeta existing = mds.get(namespace);
+          if (existing != null) {
+            return existing;
+          }
+          mds.create(metadata);
+          return null;
+        }
+      }, apps.get());
+  }
+
+  @Override
+  public void update(final NamespaceMeta metadata) {
+    Preconditions.checkArgument(metadata != null, "Namespace metadata cannot be null.");
+    appsTx.get().executeUnchecked(
+      new TransactionExecutor.Function<NamespaceMetadataStore, Void>() {
+        @Override
+        public Void apply(NamespaceMetadataStore mds) throws Exception {
+          NamespaceMeta existing = mds.get(Id.Namespace.from(metadata.getName()));
+          if (existing != null) {
+            mds.create(metadata);
+          }
+          return null;
+        }
+      }, apps.get());
+  }
+
+  @Override
+  @Nullable
+  public NamespaceMeta get(final Id.Namespace id) {
```

should it throw ``NamespaceNotFoundException`` instead of returning ``null``?